src/server/raft_client.rs

```diff
@@ -113,85 +153,84 @@ pub struct RaftClient {
+    async_executor: TaskExecutor,
 }
 
 impl RaftClient {
     pub fn new(
         env: Arc<Environment>,
         cfg: Arc<Config>,
         security_mgr: Arc<SecurityManager>,
+        grpc_thread_load: Arc<ThreadLoad>,
+        async_executor: TaskExecutor,
     ) -> RaftClient {
         RaftClient {
             env,
             conns: HashMap::default(),
             addrs: HashMap::default(),
             cfg,
             security_mgr,
+            grpc_thread_load,
+            async_executor,
         }
     }
 
     fn get_conn(&mut self, addr: &str, region_id: u64, store_id: u64) -> &mut Conn {
         let index = region_id as usize % self.cfg.grpc_raft_conn_num;
-        let cfg = &self.cfg;
-        let security_mgr = &self.security_mgr;
-        let env = &self.env;
-        // TODO: avoid to_owned
-        self.conns
-            .entry((addr.to_owned(), index))
-            .or_insert_with(|| Conn::new(Arc::clone(env), addr, cfg, security_mgr, store_id))
+        match self.conns.entry((addr.to_owned(), index)) {
+            HashMapEntry::Occupied(e) => e.into_mut(),
+            HashMapEntry::Vacant(e) => {
+                let conn = Conn::new(
+                    Arc::clone(&self.env),
+                    addr,
+                    &self.cfg,
+                    &self.security_mgr,
+                    store_id,
+                );
+                e.insert(conn)
+            }
+        }
     }
 
     pub fn send(&mut self, store_id: u64, addr: &str, msg: RaftMessage) -> Result<()> {
-        let conn = self.get_conn(addr, msg.region_id, store_id);
-        conn.buffer
-            .as_mut()
-            .unwrap()
-            .push((msg, WriteFlags::default().buffer_hint(true)));
+        if let Err(SendError(msg)) = self
+            .get_conn(addr, msg.region_id, store_id)
+            .stream
+            .send(msg)
+        {
+            error!("RaftClient send fail");
+            let index = msg.region_id as usize % self.cfg.grpc_raft_conn_num;
+            self.conns.remove(&(addr.to_owned(), index));
+
+            if let Some(current_addr) = self.addrs.remove(&store_id) {
+                if current_addr != *addr {
+                    self.addrs.insert(store_id, current_addr);
+                }
+            }
+        }
         Ok(())
     }
 
     pub fn flush(&mut self) {
-        let addrs = &mut self.addrs;
-        let mut counter: u64 = 0;
-        self.conns.retain(|&(ref addr, _), conn| {
-            let store_id = conn.store_id;
-            if !conn.alive.load(Ordering::SeqCst) {
-                if let Some(addr_current) = addrs.remove(&store_id) {
-                    if addr_current != *addr {
-                        addrs.insert(store_id, addr_current);
-                    }
+        let mut counter = 0;
+        for conn in self.conns.values_mut() {
+            if let Some(notifier) = conn.stream.get_notifier() {
+                if self.grpc_thread_load.in_heavy_load() {
```

Comment:
em, is it proper to delay the raft messages?

Reply:
Here delay raft messages **at most** 2ms. And in fact, there should be very many messages so that they can make grpc threads get in heavy load, in which case it's very easy to collect `8` messages and send them immediately.